### PR TITLE
7128 - React document listener memory leak fix.

### PR DESCRIPTION
**Pull request by mP1**
@a2ndrade

*Before* submitting a pull request, please make sure the following is done...

TODO 1. Fork the repo and create your branch from `master`.
CANT 2. If you've added code that should be tested, add tests!
NA 3. If you've changed APIs, update the documentation.
PASS 4. Ensure the test suite passes (`grunt test`).
PASS 5. Make sure your code lints (`grunt lint`) - we've done our best to make sure these rules match our internal linting guidelines.
TODO 6. If you haven't already, complete the [CLA](https://code.facebook.com/cla).

- ReactMount: calling unmountComponentFromNode on last comp,
  resets document, and removes previously added event listeners.
- ReactDOMComponent: tells ReactEventListener of any document event
  listeners.
- ReactBrowserEventEmitter: tells ReactEventListener to callback
  when document is resetted.

### CODE REVIEW

**Comment by mP1** [comment]
*Antonio*

Snapshots 3 -> 5 in both cases dynamically add & remove a task. Notice the memory remains constant.

ae/sail-client auto unmounts the "root" container when the iframe is removed.

I couldnt think of any way to write any sort of "automated test" :(

image

https://github.com/facebook/react/blob/master/CONTRIBUTING.md

Im not sure who should sign/approve the company CLA ? I havent yet tried to fill in the personal CLA. Which should i do ?

File: src/renderers/dom/client/ReactMount.js [hunk 1]
```diff
@@ -724,6 +729,15 @@ var ReactMount = {
...
   },
+  /**
+   * Rather than count mounts and unmounts, we query the dom for the DOMProperty.ROOT_ATTRIBUTE_NAME. This is used to
+   * dispose of global event listeners, when the last component is removed/unmounted.
+   * @returns Returns the number of react tagged root elements.
+   * @private
+   */
+  _countReactRootElements: function() {
+    return document.querySelectorAll('[ROOT_ATTR_NAME]').length;
+  },
```
Comment (mP1) [inline src/renderers/dom/client/ReactMount.js]: This is the only bit i might need to fix depending on which browsers are supported etc. WIll let the react guys complain.

File: src/renderers/dom/client/ReactEventListener.js [hunk 2]
```diff
@@ -172,6 +183,40 @@ var ReactEventListener = {
...
+  },
+
+  /**
+   * Removes all previously added document or global event listeners.
+   * @param {object} The document holding the last component.
+   */
+  resetDocument: function(doc) {
+    for (;;) {
+      var remover = documentEventListenerRemovers.pop();
+      if (!remover) {
```
Comment (a2ndrade) [inline src/renderers/dom/client/ReactEventListener.js]: when will `remover` be `null/undefined`? .. if we want to do it.. the code should ideally look like:
```javascript
documentEventListenerRemovers.filter(o=>!!o).forEach(remover => remover());
```

Comment (mP1) [inline src/renderers/dom/client/ReactEventListener.js]: the list always holds non null elements, pop returns null when it becomes empty.

Comment (a2ndrade) [inline src/renderers/dom/client/ReactEventListener.js]: unless there's no `filter` / `forEach` use in React's code base (i.e. if they are targetting older browsers)?? we should verify that

**Pull request closed without merging**